README.md
# Object_detection_using_YOLO_V7_For_Visually_impared_Minor_Project
Me and my teammate PIYUSH are working on a minor project. Our project title is " Object Detection Using Yolo V7 for the Visually Impaired". In this project, our main work is to create an application that will speak out 🗣️🗣️the detected Object along with its count . It has both live camera access and an image upload option for object detection.        

How do you run code in Visual Studio Code?

NOTE :: You have to download files and try to make two folder :

   (a) mpg123-1.32.3-x86-64 : It is used for the sound. 
   
   (b) Object Detection  : It is used for the code.


1) Download the file other than 
" This file is for mpg123 " in zip format.
and the 
" This file is for mpg123" Make another folder and try to install with the name of the folder "mpg123-1.32.3-x86-64"

2) Unzip the file other than "
This file is for mpg123" and try to open in vs code.

3) Open the Edit the system Environment variable from settings of your PC for "mpg123-1.32.3-x86-64"

4) Open Environment Variables
   
   
   <img width="301" alt="image" src="https://github.com/Apoorva-Dayal/Object_detection_using_YOLO_V7_For_Visually_impared_Minor_Project/assets/76716477/701ab613-b418-40f9-bcdb-e98e1e2c2265">


5) Click on Path
   
   <img width="439" alt="image" src="https://github.com/Apoorva-Dayal/Object_detection_using_YOLO_V7_For_Visually_impared_Minor_Project/assets/76716477/e0e75402-5ce5-499b-816c-26251d28e32f">


6) Add mpg123 file from the system and copy it and paste it to the path.
   
   
   <img width="374" alt="image" src="https://github.com/Apoorva-Dayal/Object_detection_using_YOLO_V7_For_Visually_impared_Minor_Project/assets/76716477/1dcaf5cd-f119-4d48-8319-e3968cda749e">

7) Now Open the vs code and open new terminal

8) Write in termial to install all packages of libraries :
   
    pip install -r requirements.txt

Now Ready!! to run the program ..


NEWS.txt
1.32.3
------
- ports/cmake: Only enable modules with GetThreadErrorMode() on Windows.
- compat: Define EOVERFLOW for ancient Windows toolchains.
- libmpg123, libsyn123: always ifdef LFS_LARGEFILE_64 (not just if)
- libsyn123: re-introduce _32 wrappers in addition to suffix-less ones
  (regression from 1.31, bug 363)

1.32.2
------

- libmpg123: Re-introduce _64 symbols on native 64 bit offset platforms. This
  was a regression since 1.31 series. Sorry, too much cleanup, not enough
  testing.
- build:
-- Better O_LARGEFILE logic, avoiding redefintion.
- ports/cmake:
-- Require C99 (bug 360, among other points, thanks to Ozkan Sezer).
-- Fix broken O_LARGEFILE logic (bug 360).
-- Typo fix and cleanup, also manual SSE switch for Android on old x86 (bug 359).

1.32.1
------
- Include man pages again in tarball and install. We cannot avoid the empty
  man directory when disabling programs with autoconf.
- Fix signal handler prototype, avoiding some justified warnings.
- ports/cmake:
-- Include CheckTypeSize, which seems to be needed sometimes (bug 357).
-- Avoid O_LARGEFILE redefinition, logic closer to autoconf.

1.32.0
------
- build:
-- Move version handling out of configure.ac to ease other build systems.
-- Include "fmt123.h" instead of <fmt123.h> in main API headers to make it more likely the
   correct one is included (at least gcc picks the one in the same directory as the
   including header first).
-- All headers are build-independent now.
-- Fix build for picky linkers by avoiding definition of wrap_getcpuflags() where it is not
   used (spurious linker error to non-exitent getcpuflags(), bug 353).
-- Handle deprecation of C99 detection macro in autoconf 2.70.
-- No use of AC_SYS_LARGEFILE anymore for explicit handling and differing choice for
   the libraries and frontend programs.
-- Added --enable-portable and --disable-largefile to configure, removing the other
   largefile-related options.
-- Added --disable-components --enable-libmpg123 to only build libmpg123 (and likewise
   --enable-libout123, --enable-libout123-modules, --enable-libsyn123) to autoconf
   build. CMake build has something  similar with BUILD_PROGRAMS and BUILD_LIBOUT123,
   which leave only libmpg123 and libsyn123 if disabled). (bug 351)
-- Consistent formatting of ./configure --help with AS_HELP_STRING().
- ports/Sony_PSP: removed
- mpg123:
-- Added --libversion.
-- Added proper A-B looping with terminal control key 'o', renamed
   --pauseloop to --presetloop.
-- Really get rid of mpg123_position() usage. (It was all lies before!)
-- Fix terminal progress info when seeking in stopped mode (1.31 regression).
-- Patch up interaction of output buffer with generic remote control, adding
   non-interruptible drain after P 3, and dropping buffer on QUIT.
-- Uppercase some generic control replies for consinstency: SILENCE, PROGRESS,
   MUTE, UNMUTE
- libmpg123, libout123, libsyn123:
-- Bumped API version for version query functions.
-- Replaced nearly all symbol renames with explicit INT123_ prefix declarations
   (intsym.h close to empty now).
- libout123:
-- Add sleep builtin output module (silent, but proper timing).
- libsyn123:
-- Introduced SYN123_PORTABLE_API for an API without off_t and ssize_t
   (see NEWS.libsyn123).
- libmpg123:
-- Internal I/O using explicit largefile support via off64_t, lseek64, fallback
   to plain 32 bit off_t.
-- Added explicit 64 bit API with 64 suffix (mpg123_tell64(), not mpg123_tell_64()).
   This allows full avoidance of ambiguus off_t. The API is always using 64 bit
   integers, regardless of internal implementation. (bug 344)
-- Introduced MPG123_PORTABLE_API for an API subset without off_t and
   ssize_t.
-- Made mpg123_seek() and friends ignore offset sign for SEEK_END (always seeking
   towards beginning, assuming negative offset) to make lseek()-conforming usage
   possible. Seeking beyond the end never made sense, so no loss of valid functionality.
- Overall use of INT123_strerror(), trying to use thread-safe strerror_l() if possible.

1.31.3
------

- build:
-- Fix --disable-8bit.
-- Fall back to generic decoder if no yasm for MSVC (bug 346).
-- Fix some pedantic compiler warnings, avoid breaking libtool wrappers.
- mpg123:
-- Fix verbose position printout for new resampling outside libmpg123 (where
   output rate differs from decoding rate).
- libsyn123:
-- Fix reconfiguration of resampler to avoid double free when reducing
   decimator stages to zero (bug 350). Thanks to Youngseok Choi for reporting
   this fuzzed issue.

1.31.2
------
- Fix build --with-network=internal only (configure logic error, bug 348).
- Fix OS/2 build with getaddrinfo() (which may support IPv6 eventually, thanks
  to Dave Yeo).

1.31.1
------
- Fix largefile aliases for the case of a largefile-insensitive build
  that still does define _FILE_OFFSET_BITS from the outside
  (sys/feature_tests.h on Illumos).

1.31.0
------

- mpg123:
-- Finally make terminal control work on Windows, for real. Building it
   was broken in 1.30.x.
-- The --control / -C switch will make mpg123 abort now if terminal
   control cannot be enabled.
-- Revert to internal network code for plain HTTP to ensure continued
   support for original shoutcast servers that do not talk proper HTTP.
   External backends are built at the same time and can be enforced using
   --network <backend>.
-- Try-witout-port for internal network code is gone. We do not need to
   keep each ancient hack for specific hosts.
-- Handle redirections independently of the backend behind net123. 
-- Set proxy environment variables when --proxy is specified, for net123
   backends to use.
-- Continue reading for long commands in generic control, avoiding
   unnecessary unfinished command errors.
-- Change error message from 'unknown command' to
   'unknown command with arguments' to avoid confusion why 'help foo'

yolo_object_detection copy 2.py
import cv2 as cv
import numpy as np
from wandb import Classes
from gtts import gTTS
import os

def text_to_speech(text):
    tts = gTTS(text=text, lang='en')
    tts.save('output.mp3')
    os.system('mpg123 output.mp3')

net = cv.dnn.readNet("yolov3.weights","yolov3.cfg")
clasees = []
with open("coco.names", 'r') as f:
    classes = [line.strip() for line in f.readlines()]

layer_name = net.getLayerNames()
output_layer = [layer_name[i - 1] for i in net.getUnconnectedOutLayers()]
colors = np.random.uniform(0, 255, size=(len(classes), 3))

img = cv.imread("sample_2.jpg") #change image according to the requirement
img = cv.resize(img, None, fx=.3, fy=.4)
height, width, channel = img.shape

blob = cv.dnn.blobFromImage(
    img, 0.00392, (416, 416), (0, 0, 0), True, crop=False)
net.setInput(blob) 
outs = net.forward(output_layer)

class_ids = []
confidences = []
boxes = []

for out in outs:
    for detection in out:
        scores = detection[5:]
        class_id = np.argmax(scores)
        confidence = scores[class_id]
        if confidence > 0.5:
            center_x = int(detection[0] * width)
            center_y = int(detection[1] * height)
            w = int(detection[2] * width)
            h = int(detection[3] * height)
            x = int(center_x - w/2)
            y = int(center_y - h/2)
            boxes.append([x, y, w, h])
            confidences.append(float(confidence))
            class_ids.append(class_id)

indexes = cv.dnn.NMSBoxes(boxes, confidences, 0.5, 0.4)

font = cv.FONT_HERSHEY_PLAIN
class_counts = {label: 0 for label in classes}

for i in range(len(boxes)):
    if i in indexes:
        x, y, w, h = boxes[i]
        label = str(classes[class_ids[i]])
        class_counts[label] += 1
        text_to_speech(f"Detected {label}")
        color = colors[class_ids[i]]
        cv.rectangle(img, (x, y), (x + w, y + h), color, 2)
        cv.putText(img, label, (x, y + 30), font, 3, color, 3)

for label, count in class_counts.items():
    if count > 0:
        print(f"Number of {label}s detected: {count}")
        text_to_speech(f"Number of {label}s detected: {count}")
  
cv.imshow("IMG", img)
cv.waitKey(10000000)


application_live.py
# main.py
from kivy.app import App
from kivy.uix.boxlayout import BoxLayout
from kivy.uix.image import Image
from kivy.uix.label import Label
from kivy.uix.button import Button
from kivy.clock import Clock
from kivy.graphics.texture import Texture

import cv2 as cv
import numpy as np
from wandb import Classes
from gtts import gTTS
import os

def text_to_speech(text):
    tts = gTTS(text=text, lang='en')
    tts.save('output.mp3')
    os.system('mpg123 output.mp3')
    # os.system('mpg123')

net = cv.dnn.readNet("yolov3.weights", "yolov3.cfg")
classes = []
with open("coco.names", 'r') as f:
    classes = [line.strip() for line in f.readlines()]

layer_names = net.getLayerNames()
output_layers = [layer_names[i - 1] for i in net.getUnconnectedOutLayers()]
colors = np.random.uniform(0, 255, size=(len(classes), 3))

class DetectionApp(App):
    
    def build(self):
        layout = BoxLayout(orientation='vertical')
        self.image = Image()
        self.start_button = Button(text='Start')
        self.start_button.bind(on_press=self.start_detection)
        self.result_label = Label()

        layout.add_widget(self.image)
        layout.add_widget(self.start_button)
        layout.add_widget(self.result_label)

        self.capture = cv.VideoCapture(0)
        self.is_detecting = False  # Initialize is_detecting to False

        Clock.schedule_interval(self.update, 1.0 / 30.0)  # Update at 30 FPS

        return layout

    def start_detection(self, instance):
        self.is_detecting = True

    def update(self, dt):
        if self.is_detecting:
            ret, frame = self.capture.read()
            frame = cv.resize(frame, None, fx=1.5, fy=1.5)
            height, width, channel = frame.shape

            blob = cv.dnn.blobFromImage(frame, 0.00392, (416, 416), (0, 0, 0), True, crop=False)
            net.setInput(blob)
            outs = net.forward(output_layers)

            class_ids = []
            confidences = []
            boxes = []

            for out in outs:
                for detection in out:
                    scores = detection[5:]
                    class_id = np.argmax(scores)
                    confidence = scores[class_id]
                    if confidence > 0.5:
                        center_x = int(detection[0] * width)
                        center_y = int(detection[1] * height)
                        w = int(detection[2] * width)
                        h = int(detection[3] * height)
                        x = int(center_x - w/2)
                        y = int(center_y - h/2)
                        boxes.append([x, y, w, h])
                        confidences.append(float(confidence))
                        class_ids.append(class_id)

            indexes = cv.dnn.NMSBoxes(boxes, confidences, 0.5, 0.4)

            font = cv.FONT_HERSHEY_PLAIN
            class_counts = {label: 0 for label in classes}

            for i in range(len(boxes)):
                if i in indexes:
                    x, y, w, h = boxes[i]
                    label = str(classes[class_ids[i]])
                    class_counts[label] += 1
                    text_to_speech(f"Detected {label}")
                    color = colors[class_ids[i]]
                    cv.rectangle(frame, (x, y), (x + w, y + h), color, 2)
                    cv.putText(frame, label, (x, y + 30), font, 3, color, 3)

            buf1 = cv.flip(frame, 0)
            buf = buf1.tostring()
            texture = Texture.create(size=(frame.shape[1], frame.shape[0]), colorfmt='bgr')
            texture.blit_buffer(buf, colorfmt='bgr', bufferfmt='ubyte')
            self.image.texture = texture

    def on_stop(self):
        self.capture.release()

if __name__ == '__main__':
    DetectionApp().run()
